File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int l = 0;
        int r = 0;
        int maxlen = 0;
        int hash[26] = {0};
        int maxfeq = 0;
        while(r < s.size()){
            hash[s[r] - 'A']++;
            maxfeq = max(maxfeq, hash[s[r] - 'A']);

            while((r-l+1) - maxfeq> k){
                hash[s[l]-'A']--;
                l++;
                maxfeq = 0;
                for(int i = 0; i < 26; i++){
                    maxfeq = max(maxfeq, hash[i]);
                }
            }

            if((r-l+1) - maxfeq <= k){
                maxlen = max(maxlen, r-l+1);
            }
            r++;

        }
        return maxlen;
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        i = 0
        j = 0
        hash_array = [0] * 26
        maxfreq = 0
        maxlen = 0
        n = len(s)

        while j < n:
            hash_array[ord(s[j]) - ord('A')] += 1
            maxfreq = max(maxfreq, hash_array[ord(s[j]) - ord('A')]) 

            while (j-i+1) - maxfreq > k:
                hash_array[ord(s[i]) - ord('A')] -= 1
                i+=1
                maxfreq = max(hash_array)
            
            maxlen = max(maxlen, j-i+1)
            j+=1
        
        return maxlen
